pkg/util/bytesbufpool/bytesbufpool.go

```diff
@@ -0,0 +1,58 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bytesbufpool
+
+import (
+	"bytes"
+	"sync"
+)
+
+var pool = sync.Pool{}
+
+const defaultCap = 100
+
+// Get a buffer with default capacity.
+func Get() *bytes.Buffer {
+	return GetWithCap(defaultCap)
+}
+
+// GetWithCap get a buffer with at least n bytes can be written to it without another allocation.
+// If n is negative, GetWithCap will panic.
+func GetWithCap(n int) *bytes.Buffer {
+	val := pool.Get()
+	if val != nil {
+		buf := val.(*bytes.Buffer)
+		if buf.Cap() < n {
+			buf.Grow(n)
+		}
+		return buf
+	}
+	return bytes.NewBuffer(make([]byte, 0, n))
+}
+
+func Put(buf *bytes.Buffer) {
+	// Proper usage of a sync.Pool requires each entry to have approximately
+	// the same memory cost. To obtain this property when the stored type
+	// contains a variably-sized buffer, we add a hard limit on the maximum
+	// buffer to place back in the pool. If the buffer is larger than the
+	// limit, we drop the buffer and not place back.
+	if buf.Cap() >= 64*1024 {
```

Comment:
IIUC `sync.pool` and the intended usage here correctly, many async clients are are filling it with buffers, and also randomly randomly pulling them off.

When (or if) are buffers purged from the pool, if there are many allocations in a short time? Would this increase observed memory usage of `istiod` (almost certainly it would, that's necessarily the tradeoff for relieving GC pressure with caching/pinning)?

Reply:
> buffers purge depend on golang GC, right? If the buffer is not Get before the next GC, it will be purge by GC, otherwise it will not be purge.

Sorry, yeah - needed to read up on `sync.pool`. It's all self-managed with an eviction queue. Basically just requires a minimum of 2 GC cycles to evict unused buffers.

Reply:
There are pathological cases where this won't perform particularly well (high concurrency+GC slow/struggling, or very weird/low `GOGC` values) but yeah otherwise this should just basically delay buffer GC for a minimum of 2 GC cycles.